libs/gl-client/src/lnurl/pay/mod.rs

```diff
@@ -0,0 +1,244 @@
+pub mod models;
+
+use anyhow::{anyhow, ensure, Result};
+use bech32::FromBase32;
+use lightning_invoice::{Invoice, InvoiceDescription};
+use models::{
+    LnUrlHttpClient, MockLnUrlHttpClient, PayRequestCallbackResponse, PayRequestResponse,
+};
+use reqwest::Url;
+use sha256;
+use std::str::from_utf8;
+use std::str::FromStr; // Import the from_utf8 function from the std::str module
+
+// Get an Invoice from a Lightning Network URL pay request
+fn parse_invoice(invoice_str: &str) -> Result<Invoice> {
+    Invoice::from_str(&invoice_str).map_err(|e| anyhow!(format!("Failed to parse invoice: {}", e)))
+}
+
+pub fn resolve_to_invoice<T: LnUrlHttpClient>(
+    http_client: T,
+    lnurl: &str,
+    amount: u64,
+) -> Result<String> {
+    let (_hrp, data, _variant) =
+        bech32::decode(lnurl).map_err(|e| anyhow!("Failed to decode lnurl: {}", e))?;
+
+    let vec = Vec::<u8>::from_base32(&data)
+        .map_err(|e| anyhow!("Failed to base32 decode data: {}", e))?;
+    let url = from_utf8(&vec).map_err(|e| anyhow!("Failed to convert to utf-8: {}", e))?;
+
+    let lnurl_pay_request: PayRequestResponse = http_client.get_pay_request_response(url)?;
+
+    validate_pay_request_response(&lnurl_pay_request, amount)?;
+
+    let m_str = lnurl_pay_request.metadata.clone();
+    let deserialized_metadata: Vec<Vec<String>> = serde_json::from_str(&m_str.to_owned())
+        .map_err(|e| anyhow!("Failed to deserialize metadata: {}", e))?;
+
+    let mut description = String::new();
+
+    for metadata in deserialized_metadata {
+        if metadata[0] == "text/plain" {
+            description = metadata[1].clone();
+        }
+    }
+
+    if description.is_empty() {
+        return Err(anyhow!("No description found"));
+    }
+
+    println!("Domain: {}", Url::parse(url).unwrap().host().unwrap());
+    println!("Description: {}", description);
+    println!(
+        "Accepted range (in millisatoshis): {} - {}",
+        lnurl_pay_request.min_sendable, lnurl_pay_request.max_sendable
+    );
+
+    let callback_url: &str =
+        &format!("{}?amount={}", lnurl_pay_request.callback.clone(), amount).to_string();
```

Comment:
heads up here, this will break if url has existing query params. common issue in lnurl implementations